realtime.go

```diff
@@ -5,60 +5,142 @@ import (
 	"context"
 	"encoding/json"
 	"errors"
+	"log/slog"
 	"net/http"
 	"strings"
 	"time"
 
 	"github.com/Flagsmith/flagsmith-go-client/v4/flagengine/environments"
 )
 
-func (c *Client) startRealtimeUpdates(ctx context.Context) {
-	err := c.UpdateEnvironment(ctx)
-	if err != nil {
-		panic("Failed to fetch the environment while configuring real-time updates")
+// realtime handles the SSE connection and reconnection logic
+type realtime struct {
+	client        *Client
+	ctx           context.Context
+	log           *slog.Logger
+	streamURL     string
+	envUpdatedAt  time.Time
+	backoff       *backoff
+	reconnectChan chan struct{}
+}
+
+// newRealtime creates a new realtime instance
+func newRealtime(client *Client, ctx context.Context, streamURL string, envUpdatedAt time.Time) *realtime {
+	return &realtime{
+		client: client,
+		ctx:    ctx,
+		log: client.log.With(
+			slog.String("worker", "realtime"),
+			slog.String("stream", streamURL),
+		),
+		streamURL:     streamURL,
+		envUpdatedAt:  envUpdatedAt,
+		backoff:       newBackoff(),
+		reconnectChan: make(chan struct{}, 1),
 	}
-	env, _ := c.environment.Load().(*environments.EnvironmentModel)
-	stream_url := c.config.realtimeBaseUrl + "sse/environments/" + env.APIKey + "/stream"
-	envUpdatedAt := env.UpdatedAt
+}
+
+// start begins the realtime connection process
+func (r *realtime) start() {
+	r.log.Debug("connecting to realtime")
+	defer func() {
+		r.log.Info("stopped")
+	}()
 	for {
 		select {
-		case <-ctx.Done():
+		case <-r.ctx.Done():
 			return
+		case <-r.reconnectChan:
+			// Reset backoff on successful reconnect
+			r.backoff.reset()
 		default:
-			resp, err := http.Get(stream_url)
-			if err != nil {
-				c.log.Errorf("Error connecting to realtime server: %v", err)
-				continue
+			if err := r.connect(); err != nil {
+				r.log.Error("failed to connect", "error", err)
+				r.wait()
 			}
-			defer resp.Body.Close()
-
-			scanner := bufio.NewScanner(resp.Body)
-			for scanner.Scan() {
-				line := scanner.Text()
-				if strings.HasPrefix(line, "data: ") {
-					parsedTime, err := parseUpdatedAtFromSSE(line)
-					if err != nil {
-						c.log.Errorf("Error reading realtime stream: %v", err)
-						continue
-					}
-					if parsedTime.After(envUpdatedAt) {
-						err = c.UpdateEnvironment(ctx)
-						if err != nil {
-							c.log.Errorf("Failed to update the environment: %v", err)
-							continue
-						}
-						env, _ := c.environment.Load().(*environments.EnvironmentModel)
-
-						envUpdatedAt = env.UpdatedAt
-					}
+		}
+	}
+}
+
+// connect establishes and maintains the SSE connection
+func (r *realtime) connect() error {
+	resp, err := http.Get(r.streamURL)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	r.log.Info("connected")
+	r.reconnectChan <- struct{}{}
```

Comment:
Is this the only place where we are sending the signal? If that's the case, I think we can get rid of the channel and just call `r.backoff.reset()` here?